server/src/main/java/org/opensearch/index/compositeindex/datacube/startree/builder/OffHeapStarTreeBuilder.java

```diff
@@ -0,0 +1,937 @@
+ * This is necessary because Lucene maintains immutability of data, and an {@code IndexOutput} cannot be kept open while creating an
+ * {@code IndexInput} on the same file, as all file contents may not be visible in the reader. Therefore, the {@code IndexOutput} must be
+ * closed to ensure all data can be read before creating an {@code IndexInput}. Additionally, an {@code IndexOutput} cannot be reopened,
+ * so a new file is created for the new star tree documents.
+ * <p>The set of 'star-tree.documents' files is maintained, and a tracker array is used to keep track of the start document ID for each file.
+ * Once the number of files reaches a set threshold, the files are merged.
+
+ @opensearch.experimental
+ **/
+
+@ExperimentalApi
+public class OffHeapStarTreeBuilder extends BaseStarTreeBuilder {
+    private static final Logger logger = LogManager.getLogger(OffHeapStarTreeBuilder.class);
+    private static final String SEGMENT_DOC_FILE_NAME = "segment.documents";
+    private static final String STAR_TREE_DOC_FILE_NAME = "star-tree.documents";
+    // TODO : Should this be via settings ?
+    private static final int DEFAULT_FILE_COUNT_MERGE_THRESHOLD = 5;
+    private final int fileCountMergeThreshold;
+    private final List<Integer> starTreeDocumentOffsets;
+    private int numReadableStarTreeDocuments;
+    final IndexOutput segmentDocsFileOutput;
+    IndexOutput starTreeDocsFileOutput;
+    IndexInput starTreeDocsFileInput;
+    IndexInput segmentDocsFileInput;
+    RandomAccessInput segmentRandomInput;
+    private RandomAccessInput starTreeDocsFileRandomInput;
+    SegmentWriteState state;
+    Map<String, Integer> fileToByteSizeMap;
+    int starTreeFileCount = -1;
+    int prevStartDocId = Integer.MAX_VALUE;
+    int currBytes = 0;
+    int docSizeInBytes = -1;
+    TrackingDirectoryWrapper tmpDirectory;
+
+    /**
+     * Builds star tree based on star tree field configuration consisting of dimensions, metrics and star tree index
+     * specific configuration.
+     *
+     * @param starTreeField holds the configuration for the star tree
+     * @param state         stores the segment write state
+     * @param mapperService helps to find the original type of the field
+     */
+    protected OffHeapStarTreeBuilder(StarTreeField starTreeField, SegmentWriteState state, MapperService mapperService) throws IOException {
+        this(starTreeField, state, mapperService, DEFAULT_FILE_COUNT_MERGE_THRESHOLD);
+    }
+
+    /**
+     * Builds star tree based on star tree field configuration consisting of dimensions, metrics and star tree index
+     * specific configuration.
+     *
+     * @param starTreeField holds the configuration for the star tree
+     * @param state         stores the segment write state
+     * @param mapperService helps to find the original type of the field
+     * @param fileThreshold threshold for number of files after which we merge the files
```

Comment:
Since it copies all files to a single file, there is some disk usage for it. 
But its not invoked often, it depends on number of dimensions and max leaf docs param. 
For 5 dims or more, this gets called, which i feel is a rare case as users will have lesser dims than that and even if there are more dims, depending on max leaf docs, there might not be equivalent levels in star tree. 

Hence set the threshold up as 5.

But to your point, we can take this as a setting as well if needed ? depending on node resources, not sure if there are good indicator for IO.

Reply:
Lets create an issue to measure the impact on IO due to a low/high value

Reply:
https://github.com/opensearch-project/OpenSearch/issues/14813#issuecomment-2261059675 - for now tracking here, will create a perf meta issue